src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -531,10 +549,10 @@ internal ODataResourceMetadataContextWithModel(ODataResourceBase resource, IODat
                         IEdmEntityType entityType = this.actualResourceType as IEdmEntityType;
                         if (entityType != null)
                         {
-                            this.keyProperties = GetPropertyValues(entityType.Key(), resource, entityType, this.requiresId).ToArray();
+                            this.keyProperties = GetPropertyValues(entityType.Key(), resource, this.actualResourceType, this.requiresId).ToArray();
                         }
 
-                        if (!ValidateEntityTypeHasKeyProperties(this.keyProperties, this.ActualResourceTypeName, this.requiresId))
+                        if (!ValidateEntityTypeHasKeyProperties(this.keyProperties, resource, entityType, this.requiresId))

```

Comment:
`entityType` can be null here...

Reply:
Good catch. I can pass the `actualResourceType` as-is (`IEdmStructuredType`) instead.

Reply:
the `ToArray()` is also not needed here anymore.